rbx_types/src/attributes/reader.rs

```diff
@@ -0,0 +1,180 @@
+// Thanks to Anaminus! https://github.com/RobloxAPI/rbxattr/blob/master/spec.md
+
+use super::*;
+use crate::{basic_types::*, brick_color::BrickColor, variant::Variant};
+use std::{
+    collections::HashMap,
+    io::{self, Read},
+};
+
+fn read_u8<R: Read>(mut reader: R) -> io::Result<u8> {
+    let mut bytes = [0u8; 1];
+    reader.read_exact(&mut bytes)?;
+    Ok(bytes[0])
+}
+
+fn read_i32<R: Read>(mut reader: R) -> io::Result<i32> {
+    let mut bytes = [0u8; 4];
+    reader.read_exact(&mut bytes)?;
+    Ok(i32::from_le_bytes(bytes))
+}
+
+fn read_u32<R: Read>(mut reader: R) -> io::Result<u32> {
+    let mut bytes = [0u8; 4];
+    reader.read_exact(&mut bytes)?;
+    Ok(u32::from_le_bytes(bytes))
+}
+
+fn read_f32<R: Read>(mut reader: R) -> io::Result<f32> {
+    let mut bytes = [0u8; 4];
+    reader.read_exact(&mut bytes)?;
+    Ok(f32::from_le_bytes(bytes))
+}
+
+fn read_string<R: Read>(mut reader: R) -> io::Result<String> {
+    let size = read_u32(&mut reader)? as usize;
+    let mut characters = vec![0u8; size];
+    reader.read_exact(&mut characters)?;
+    Ok(characters.into_iter().map(|byte| byte as char).collect())
+}
+
+fn read_color3<R: Read>(mut reader: R) -> io::Result<Color3> {
+    Ok(Color3::new(
+        read_f32(&mut reader)?,
+        read_f32(&mut reader)?,
+        read_f32(&mut reader)?,
+    ))
+}
+
+fn read_udim<R: Read>(mut reader: R) -> io::Result<UDim> {
+    Ok(UDim::new(read_f32(&mut reader)?, read_i32(&mut reader)?))
+}
+
+fn read_vector2<R: Read>(mut reader: R) -> io::Result<Vector2> {
+    Ok(Vector2::new(read_f32(&mut reader)?, read_f32(&mut reader)?))
+}
+
+/// Reads through an attribute property (AttributesSerialize) and returns a map of attribute names -> values.
+pub fn get_attributes<R: Read>(mut value: R) -> Result<HashMap<String, Variant>, AttributeError> {
+    let size = read_u32(&mut value).map_err(|_| AttributeError::InvalidSize)?;
+    let mut attributes = HashMap::with_capacity(size as usize);
+
+    for _ in 0..size {
+        let name = read_string(&mut value).map_err(|_| AttributeError::InvalidName)?;
+
+        let attribute_key =
```

Comment:
We can probably just cast directly into the Enum using `try_into` like we do in rbx_binary:
https://github.com/rojo-rbx/rbx-dom/blob/master/rbx_binary/src/deserializer.rs#L419

Reply:
Enum as in the u8? That's what it looks like in your example.

Reply:
Upon a closer look at the implementation in rbx_binary, it seems as though I made an assumption I shouldn't have. In my defense, the code was written before I touched the codebase, but I have no one to blame but myself for not reading it.

I was mostly hoping to reduce some boilerplate but it seems to just be a hidden abstraction in rbx_binary. 🙁